Terraform Cloud per-run variables (#1978)
hashicorp/terraform-website: content/source/docs/cloud/workspaces/variables.html.md

If a required input variable is missing, Terraform plans in the workspace will fail and print an explanation in the log.

## Scope

Each environment and Terraform variable can have one of the following scopes:

| Scope               | Description                                                                        | Resources                                                                                                                                                                                                                                                                                                      |
|---------------------|------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | Run-Specific        | Apply to a specific run within a single workspace                                  | [Run-Specific Variables](/docs/cloud/workspaces/managing-variables.html#run-specific-variables)                                                                                                                                                                                                                |

Author: I ended up using the word Specify because we really don't create entities by using run-specific variables

If a required input variable is missing, Terraform plans in the workspace will fail and print an explanation in the log.

> #### Loading Variables from the Command Line

Author: I like this idea! I came away thinking that I may not have used a critical phrasing to identify the feature.

When using Terraform 1.1 or later, you can set input variable values for a particular plan/apply run on the command line using either of the following mechanisms:

* Specifying `-var` and `-var-file` arguments
* Local environment variables named `TF_VAR_` followed by the name of a declared variable

> Variable values set in either of these ways will take precedence over Workspace-Specific values. See the [Precedence](#precedence) section for more details. For more information about specifying CLI input variable values, see the Terraform documentation for [Variables on the Command Line](/docs/language/values/variables.html#variables-on-the-command-line) or for the [Input Variables on the Command Line](/docs/cli/commands/plan.html#input-variables-on-the-command-line) for the plan command.

Reviewer: So, I think that you can only have TFC pick up a `terraform.tfvars` file if you do this command line flag, right? It doesn't work the same way as `auto.tfvars`, right? Like, if I try to to provide TFC with a `terraform.tfvars` file the same way I would with `auto.tfvars`, does it recognize it? We should probably be explicit about if that does or does not work and why.

```suggestion
Variable values set in either of these ways will take precedence over Workspace-Specific values. See the [Precedence](#precedence) section for more details. For more information about specifying CLI input variable values, see the Terraform documentation for [Variables on the Command Line](/docs/language/values/variables.html#variables-on-the-command-line) or for the [Input Variables on the Command Line](/docs/cli/commands/plan.html#input-variables-on-the-command-line) for the plan command.
```

Author: TFC will actually overwrite your terraform.tfvars file with one of it's own, so those values are completely ignored by TFC. For some reason, I thought we documented that all over the place but I don't see it anywhere in the doc! I'll add it.

Terraform Cloud prioritizes and overwrites conflicting variables according to the following precedence:

> ### 1. Command line Argument Variables

Author: I don't want to get "Variables Set" mixed up with "Variable Sets" -- also via is a latin word, right?

> ### 1. Command line Argument Variables

Reviewer: We use via in other places throughout the docs, but that's a fair point. I also see your point about leaving this as is so that two things don't start with "Variables" - thank you!

When using a CLI workflow, variables applied to a run with either `-var` or `-var-file` overwrite workspace-specific and variable set variables that have the same key.

### 2. Local Environment Variables Prefixed with `TF_VAR_`

> When using a CLI workflow, local environment variables prefixed with `TF_VAR_` (e.g., `TF_VAR_key`) overwrite workspace-specific, variable set, and `.auto.tfvars` file variables that have the same key.

Reviewer: i read it again and realized key might be confusing because we keep talking about variables with the same key elsewhere. 

```suggestion
When using a CLI workflow, local environment variables prefixed with `TF_VAR_` (e.g., `TF_VAR_replicas`) overwrite workspace-specific, variable set, and `.auto.tfvars` file variables that have the same key.
```

### 5. Global Variable Sets

Non-global variable sets always take precedence over global variable sets that are applied to all workspaces within an organization. This is true regardless of when the global variable set was applied.

### 6. `auto.tfvars` Variable Files

Variables in the Terraform Cloud workspace and variables provided through the command line always overwrite variables with the same key from files ending in `auto.tfvars`.

### Precedence Example

Consider an example workspace that has the following variables applied:

| Source                | Scope              | Date Applied | ACCESS_KEY | ACCESS_ID  | VAR1 | KEY1 | VAR2 | replicas |
|-----------------------|--------------------|--------------|------------|------------|------|------|------|----------|

> | Command line argument | Run-Specific       |              |            |            |      |      |      | `9`      |

Reviewer: Everything else is caps so it looks weird if this isn't imo

```suggestion
| Command Line Argument | Run-Specific       |              |            |            |      |      |      | `9`      |
```

> | Name           | Scope              | Date Applied | ACCESS_KEY | ACCESS_ID  | VAR1 | KEY1 | VAR2 | foo |
> |----------------|--------------------|--------------|------------|------------|------|------|------|-----|
> | TF_VAR_foo=8   | Run-Specific       |              |            |            |      |      |      | `8` |
> | -var="foo=9"   | Run-Specific       |              |            |            |      |      |      | `9` |
> | Variables      | Workspace-Specific | 10/1         | `g47fh474` | `874hf7u4` | `h`  |      |      | `1` |
> | Variable Set A | Non-Global         | 10/4         |            |            | `y`  | `x`  |      | `2` |
> | Variable Set B | Global             | 10/20        |            |            |      | `z`  | `a`  | `3` |

Reviewer: Okay so, I was thinking about this example and about how to make it possibly clearer. One thing is that I want to change name to Source because I think maybe it'll be clearer where these things are coming from (and I tried to use language that would match the precedence headings above) see what you think about that. And THEN I was thinking that maybe it'd be nice to use a real variable name instead of just "foo". We have a Learn tutorial where we are setting up a kubernetes deployment and one of the properties we have to set is the number of replicas: https://learn.hashicorp.com/tutorials/terraform/cdktf-applications

Anyway, I also updated the example a bit with that in mind. Let me know what you think of that, too and/or if we should use a different example variable :-) 
```suggestion
| Source            | Scope              | Date Applied | ACCESS_KEY | ACCESS_ID  | VAR1 | KEY1 | VAR2 | REPLICAS |
| ----------------- | ------------------ | ------------ | ---------- | ---------- | ---- | ---- | ---- | -------- |
| Command Line      | Run-Specific       | 10/23        |            |            |      |      |      | `4`      |
| Local Environment | Run-Specific       | 10/23        |            |            |      |      |      | `3`      |
| Workspace UI      | Workspace-Specific | 10/1         | `g47fh474` | `874hf7u4` | `h`  |      |      | `1`      |
| Variable Set A    | Non-Global         | 10/4         |            |            | `y`  | `x`  |      | `2`      |
| Variable Set B    | Global             | 10/20        |            |            |      | `z`  | `a`  | `3`      |
```

Author: Sounds good!

> When you trigger a run, Terraform Cloud applies the following variables:

Reviewer: ```suggestion
When you trigger a run through the command line, Terraform Cloud applies the following variables:
```

> - **Run-Specific:** `foo` is specified in all scopes. The -var value takes precedence over all other values, including the run-specific TF_VAR_foo values.

Reviewer: ```suggestion
- **Run-Specific:** `REPLICAS` from the command line. That means `REPLICAS` equals `4` for this run. Variables set via the command line take precedence over all other values, including the run-specific `TF_VAR` value set in your local environment.
```

> - **Run-Specific:** `replicas` from the command line. That means `replicas` equals `9` for this run. Variables set via the command line take precedence over all other values, including the run-specific `TF_VAR_replicas` value set in your local environment.

Reviewer: ```suggestion
- **Run-Specific:** `replicas` from the command line. That means `replicas` equals `9` for this run. Variables set from the command line take precedence over all other values, including the run-specific `TF_VAR_replicas` value set in your local environment.
```

- **Workspace-Specific:** `ACCESS_KEY`, `ACCESS_ID`, and `VAR1`. That means `VAR1` equals `h` for this run, overwriting the value in Variable Set A.

- **Variable Set A:** `KEY1`. That means `KEY1` equals `x` for this run, overwriting the value in Variable Set B.

- **Variable Set B:** `VAR2`. This is a global variable set, so Variable Set A takes precedence regardless of when it was applied.

> ![An example scenario demonstrating variable precedence in Terraform Cloud](/docs/cloud/workspaces/images/variable-precedence-example.png)

Reviewer: I am going to send you a new image we can use on slack! Please replace the current one with whatever the final visual is (once we settle on the new variable names, etc.)

Terraform Cloud prioritizes and overwrites conflicting variables according to the following precedence:

### 1. Command line Argument Variables

> When using a CLI workflow, variables specified using `-var` or `-var-file` during plan/apply will overwrite all other variable values.

Author: Yes they overwrite the same key which I clarified- thanks!

### 2. TF_VAR_ Prefixed Environment Variables

> When using a CLI workflow, variables found in the local environment that are prefixed TF_VAR_ followed by the name of a declared variable overwrite workspace-specific, variable set values, and variable file values.

Reviewer: Just checking -  I have a `TF_VAR` variable declared in my environment, it overrides any variables in `auto.tf.vars` but NOT run-specific things from the command line, right? I wonder if we should be more specific about "file values" and specifically say that it overwrites `auto.tfvars`

Consider an example workspace that has the following variables applied:

> | Name           | Scope              | Date Applied | ACCESS_KEY | ACCESS_ID  | VAR1 | KEY1 | VAR2 |

Author: I had originally excluded it because I would need to add a new example variable to the table, but I think it's a good tradeoff. Let me know what you think.
